lightning/src/ln/functional_tests.rs

```diff
@@ -9913,3 +9915,46 @@ fn test_max_dust_htlc_exposure() {
 	do_test_max_dust_htlc_exposure(false, ExposureEvent::AtUpdateFeeOutbound, false);
 	do_test_max_dust_htlc_exposure(false, ExposureEvent::AtUpdateFeeOutbound, true);
 }
+
+#[test]
+fn test_mpp_receive_timeout() {
+	let chanmon_cfgs = create_chanmon_cfgs(2);
+	let node_cfgs = create_node_cfgs(2, &chanmon_cfgs);
+	let node_chanmgrs = create_node_chanmgrs(2, &node_cfgs, &[None, None]);
+	let nodes = create_network(2, &node_cfgs, &node_chanmgrs);
+
+	let chan_id = create_announced_chan_between_nodes(&nodes, 0, 1, InitFeatures::known(), InitFeatures::known()).0.contents.short_channel_id;
+
+	let (mut route, payment_hash, _payment_preimage, payment_secret) = get_route_and_payment_hash!(&nodes[0], nodes[1], 100000);
+	let path = route.paths[0].clone();
+	route.paths.push(path);
+	route.paths[0][0].pubkey = nodes[1].node.get_our_node_id();
+	route.paths[0][0].short_channel_id = chan_id;
+
+	let cur_height = CHAN_CONFIRM_DEPTH + 1; // route_payment calls send_payment, which adds 1 to the current height. So we do the same here to match.
+	let payment_id = PaymentId([42; 32]);
+
+	nodes[0].node.send_payment_along_path(&route.paths[0], &route.payment_params, &payment_hash, &Some(payment_secret), 200000, cur_height, payment_id, &None).unwrap();
+	check_added_monitors!(nodes[0], 1);
+	let mut events = nodes[0].node.get_and_clear_pending_msg_events();
+	assert_eq!(events.len(), 1);
+	
+	// Now do the relevant commitment_signed/RAA dances along the path, noting that the final
+	// hop should *not* yet generate any PaymentReceived event(s).
+	pass_along_path(&nodes[0], &[&nodes[1]], 100000, payment_hash, Some(payment_secret), events.drain(..).next().unwrap(), false, None);
+	
```

Comment:
nit: you have spare whitespace here on a blank line, and on a few of the lines further down.